auth/src/main/java/com/firebase/ui/auth/data/remote/AnonymousSignInHandler.java

```diff
@@ -0,0 +1,78 @@
+package com.firebase.ui.auth.data.remote;
+
+import android.app.Application;
+import android.content.Intent;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.annotation.RestrictTo;
+import android.support.annotation.VisibleForTesting;
+
+import com.firebase.ui.auth.AuthUI;
+import com.firebase.ui.auth.IdpResponse;
+import com.firebase.ui.auth.data.model.FlowParameters;
+import com.firebase.ui.auth.data.model.Resource;
+import com.firebase.ui.auth.data.model.User;
+import com.firebase.ui.auth.ui.HelperActivityBase;
+import com.firebase.ui.auth.viewmodel.ProviderSignInBase;
+import com.google.android.gms.tasks.OnFailureListener;
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.firebase.FirebaseApp;
+import com.google.firebase.auth.AuthResult;
+import com.google.firebase.auth.FirebaseAuth;
+
+
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+public class AnonymousSignInHandler extends ProviderSignInBase<FlowParameters> {
+
+    @VisibleForTesting
+    public FirebaseAuth mAuth;
+
+    public AnonymousSignInHandler(Application application) {
+        super(application);
+    }
+
+    @Override
+    protected void onCreate() {
+        mAuth = getAuth();
+    }
+
+    @Override
+    public void startSignIn(@NonNull HelperActivityBase activity) {
+        setResult(Resource.<IdpResponse>forLoading());
+
+        // Calling signInAnonymously() will always return the same anonymous user if already
+        // available. This is enforced by the client SDK.
+        mAuth.signInAnonymously()
+                .addOnSuccessListener(new OnSuccessListener<AuthResult>() {
+                    @Override
+                    public void onSuccess(AuthResult result) {
+                        setResult(Resource.<IdpResponse>forSuccess(initResponse(
+                                result.getAdditionalUserInfo().isNewUser())));
+                    }
+                })
+                .addOnFailureListener(new OnFailureListener() {
+                    @Override
+                    public void onFailure(@NonNull Exception e) {
+                        setResult(Resource.<IdpResponse>forFailure(e));
+                    }
+                });
+
+    }
+
+    @Override
+    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {}
+
+    private IdpResponse initResponse(boolean isNewUser) {
+        return new IdpResponse.Builder(
+                new User.Builder(AuthUI.AnonymousAuthProvider.PROVIDER_ID, null)
+                        .build())
+                .setNewUser(isNewUser)
+                .build();
+    }
+
+    // TODO: Change ProviderSignInBase to extend from AuthViewModelBase
```

Comment:
`ProviderSignInBase` classes are meant to only retrieve remote provider data, hence making it difficult to get an auth instance as you've noticed. It was also supposed to be UI focused which is why there's a bunch of activity related stuff.

I feel like we should make this class be a bare bones middle man that sends an `IdpResponse` to an `AnonymousProviderResponseHandler` with the auth stuff. @samtstern What do you think?

Reply:
@SUPERCILEX in this version which class actually would call `signInAnonymously`?

Reply:
`signInAnonymously` would be called in the `AnonymousProviderResponseHandler`. And `AnonymousSignInHandler` would be something as simple as:
```java
// ...

@Override
public void startSignIn(@NonNull HelperActivityBase activity) {
    setResult(Resource.<IdpResponse>forSuccess(new IdpResponse.Builder(
                new User.Builder(AuthUI.AnonymousAuthProvider.PROVIDER_ID, null).build())
                .build()));
}
```

Not sure if that's actually better though since having `AnonymousSignInHandler` becomes kinda pointless. I'm just not a huge fan of doing auth stuff in a non-auth view model. 🤷‍♂️

Reply:
@samtstern thoughts?

Reply:
Maybe we compromise and **only** have the `ResponseHandler` but call it immediately, rather than having the useless `ViewModel`.  I agree we shouldn't do the signing in in the non-auth view model if we can avoid it.